### DISCUSSION TITLE
Testing lexer without a parser?

### DISCUSSION THREAD

**Discussion by osa1** [tree-sitter/tree-sitter#1036] [Q&A] [answered]
I'm looking for a way to test lexers without parsing. For example, perhaps a `tree-sitter` command, when pointed to .so for a lexer, could tokenize a given file and dump the list of tokens, similar to `tree-sitter parse`. Looking at the documentation I can't see anything relevant. Is there a way to do this currently?

**Comment by ahlinc** [comment]
I suppose it isn't be possible at all cause parsing and lexing are parallel processes in the tree-sitter core, this is the way how an incremental parsing works in it.

**Reply by osa1** [reply]
I don't think there's an algorithmic problem here, given a lexer for incremental parsing, you could tokenize an entire file by just asking for the next token after previous token (surely there must be some way to get the first token in a source, use that for starting lexing). I think the main question is what to pass for the "valid symbols" argument. I think for this mode it would make sense include all token types in "scan" calls.

**Reply by ahlinc** [reply]
There are many grammars that can't be lexed correctly without a context provided by the parser and heredoc delimiters in the bash grammar is a good example.

Could you explain your purpose to test the lexing separately?

If you are looking for a list of all possible tokens in some grammar you can find them in the `node-types.json` file that is [generated](https://github.com/tree-sitter/tree-sitter-c/blob/master/src/node-types.json) by `tree-sitter generate` command for every grammar.

**Comment by maxbrunsfeld** [comment, marked as answer]
Tree-sitter's generated lexers are very coupled to the parsers. For every token, you need to know:
* whether or not any external tokens are valid at the current position and if so, what `valid_tokens` should be passed to the external scanner function.
* for the regular scanner function, what lex state number you should start in. This is determined by what tokens are valid

Lexing won't give the correct results if that data is not known. I think that if it were absolutely necessary to avoid some of the performance cost of parsing, it would be possible to perform some special parsing *mode* where the Tree is discarded, but the states are still tracked. I don't know if it is worth the complexity though; it'd be a pretty invasive change to the runtime library for a pretty niche use-case.

**Comment by agirardeau** [comment]
Is there a way to view only the output tokens from the lexer? Just because the parser has to run doesn't mean the lexer output can't be displayed for debugging.

**Reply by ahlinc** [reply]
@agirardeau you can try to grep by the tree-sitter debug output like:
```bash
tree-sitter parse src/tree_sitter/parser.h -d |& grep -P 'lex.*sym:'
```
